src/display-x11.cc

```diff
@@ -679,6 +423,314 @@ bool display_output_x11::main_loop_wait(double t) {
+    if (cursor_over_conky) {
+      if (!cursor_inside) {
+        llua_mouse_hook(mouse_crossing_event(
+            mouse_event_t::AREA_ENTER, data->root_x - window.x,
+            data->root_y - window.x, data->root_x, data->root_y));
+      }
+      cursor_inside = true;
+    } else if (cursor_inside) {
+      llua_mouse_hook(mouse_crossing_event(
+          mouse_event_t::AREA_LEAVE, data->root_x - window.x,
+          data->root_y - window.x, data->root_x, data->root_y));
+      cursor_inside = false;
+    }
+  }
+  XFreeEventData(display, &ev.xcookie);
+  return true;
+}
+#endif /* BUILD_XINPUT */
+EV_HANDLER(mouse_input) {
+  if (ev.type != ButtonPress && ev.type != ButtonRelease &&
+      ev.type != MotionNotify)
+    return false;
+  if (ev.xany.window != window.window) return true;  // Skip other windows
+
+#ifdef BUILD_MOUSE_EVENTS
+  switch (ev.type) {
+    case ButtonPress: {
+      modifier_state_t mods = x11_modifier_state(ev.xbutton.state);
+      if (ev.xbutton.button >= 4 &&
+          ev.xbutton.button <= 7) {  // scroll "buttons"
+        scroll_direction_t direction = x11_scroll_direction(ev.xbutton.button);
+        *consumed = llua_mouse_hook(
+            mouse_scroll_event(ev.xbutton.x, ev.xbutton.y, ev.xbutton.x_root,
+                               ev.xbutton.y_root, direction, mods));
+      } else {
+        mouse_button_t button = x11_mouse_button_code(ev.xbutton.button);
+        *consumed = llua_mouse_hook(mouse_button_event(
+            mouse_event_t::MOUSE_PRESS, ev.xbutton.x, ev.xbutton.y,
+            ev.xbutton.x_root, ev.xbutton.y_root, button, mods));
+      }
+    }
+    case ButtonRelease: {
+      /* don't report scroll release events */
+      if (ev.xbutton.button >= 4 && ev.xbutton.button <= 7) return true;
+
+      modifier_state_t mods = x11_modifier_state(ev.xbutton.state);
+      mouse_button_t button = x11_mouse_button_code(ev.xbutton.button);
+      *consumed = llua_mouse_hook(mouse_button_event(
+          mouse_event_t::MOUSE_RELEASE, ev.xbutton.x, ev.xbutton.y,
+          ev.xbutton.x_root, ev.xbutton.y_root, button, mods));
+    }
+    case MotionNotify: {
+      modifier_state_t mods = x11_modifier_state(ev.xmotion.state);
+      *consumed = llua_mouse_hook(mouse_move_event(ev.xmotion.x, ev.xmotion.y,
+                                                   ev.xmotion.x_root,
+                                                   ev.xmotion.y_root, mods));
+    }
+  }
+#else  /* BUILD_MOUSE_EVENTS */
+  // always propagate mouse input if not handling mouse events
+  *consumed = false;
+#endif /* BUILD_MOUSE_EVENTS */
+
+  if (!own_window.get(*state)) return true;
+  switch (own_window_type.get(*state)) {
+    case window_type::TYPE_NORMAL:
+    case window_type::TYPE_UTILITY:
+      // decorated normal windows always consume events
+      if (!TEST_HINT(own_window_hints.get(*state), HINT_UNDECORATED)) {
+        *consumed = true;
```

Comment:
So now, it still passes the events to Lua, but force-disables propagation in case of normal decorated window, or desktop. Also added comments to make it clear in future.